enhancements/agent-installer/automated-workflow-for-agent-based-installer.md

```diff
@@ -0,0 +1,624 @@
+  the mirrored OpenShift image is left to a future enhancement.
+- Describing a mechanism to add and remove nodes from the cluster after
+  it is installed is left to a future enhancement.
+- Generating image formats other than ISO is left to a future
+  enhancement.
+- Booting hosts via PXE is left to a future enhancement.
+- Creating minimal ISO images with a separate rootfs is left to a
+  future enhancement.
+- We are not currently considering replacing the IPI installer or
+  assisted installer SaaS.
+- Describing an end-to-end solution for users without their own
+  orchestration tool is left to a future enhancement.
+- Collecting telemetry data for connected deployments is left to a
+  future enhancement.
+- Details of securing the communication between agents and services to
+  ensure only desired hosts join the cluster are left to a future
+  enhancement.
+- Details of a workflow for deploying in settings with DHCP assignment
+  of IP addresses is left to a future enhancement.
+- It is not currently a requirement that the downloadable form of the
+  installer support installing multiple versions of OpenShift. We will
+  address the underlying requirements behind that request in other
+  ways, which will be discussed elsewhere.
+
+## Proposal
+
+We propose to extend the `openshift-install` binary with a new
+sub-command, `agent create-image`, for generating a bootable image
+containing all of the information needed to deploy a cluster.  A
+command line tool avoids issues with prerequisites for deploying a
+long-running service, and can be run on a low-resource host such as a
+laptop or even on a host at a different site before deployment begins.
+
+Most enterprise grade hardware includes a management controller that
+can present an ISO image to a host virtually, allowing the host to
+boot from the image remotely. Hypervisors such as kvm and vSphere also
+support booting from ISOs. Therefore, we will concentrate on that
+format and approach first. Building minimal ISOs with separate root
+filesystems or images that can be used with hosts that PXE boot will
+come later.
+
+We propose to leave up to the user's orchestration tool the
+responsibilities of managing and storing the images, serving them on
+the network, and especially configuring hosts to boot them. This gives
+users the flexibility to perform those operations in any way they
+like, with whatever tool they prefer. A design for an all encompassing
+approach based on tools provided entirely by Red Hat is left for a
+later enhancement.
+
+The `agent create-image` command will need file-based inputs to
+describe the cluster and node-specific configuration. Using the
+existing file and custom resource types allows us to use the same
+version management techniques that we do for the APIs within a
+cluster.  We will use the `install-config.yaml` format for
+cluster-wide settings. We will use a new file, tentatively named
+`agent-config.yaml`, for settings that apply to only agent-based
```

Comment:
I asked the same question and haven't received an answer.

Reply:
To me that conflicts with the goal above:

> - We want to avoid inventing completely new ways to describe OpenShift
  clusters. Therefore we will reuse existing inputs to the installer
  and to tools such as the zero-touch provisioning (ZTP) system in
  multi-cluster engine (MCE).

It also seems very weird to me that we expect users to install cluster0 using these APIs and any additional clusters with ZTP APIs.

Reply:
ZTP APIs are still there.

Don't forget that this enhancement is eventually going to be followed by one for an interactive workflow where the data from agent-config.yaml is still required (because it's needed to boot the hosts at all), but none of the data from the ZTP manifest or the install-config.yaml is, because the user will enter that through the assisted UI.

Reply:
If there's a gap where the CRDs that make up the ZTP approach don't have those features, because those features are (at least for now) unique to billi, it's a reasonable option to add them as annotations rather than create a new file with its own API. Was that considered and/or ruled out?

Reply:
They'll be needed even if there's _no_ ZTP manifests, because you're doing an interactive install. You can't have an annotation on nothing.

Reply:
I guess I'm not following, so maybe another level of detail would make things more clear. Forgive me for mostly stating the obvious.

For on-prem agent-based provisioning with ACM, we would create in InfraEnv, a ClusterDeployment, and an AgentClusterInstall resource. When the UI is involved, it uses a wizard to collect the info that goes in those resources. Those resources are always required to do agent-based installation with ACM.

IIUC, here you're trying to provide a way for a user of this CLI to pass essentially the same information, plus a few additional items that you listed above, which aren't yet covered by CRDs. That user will put the information into one or more text files and then run the CLI.

This document proposes to use the install-config and a new agent-config as the structure for collecting the info in text files. I'm trying to ask, what if we used the CRDs instead, and filled in any small gaps with annotations? Or we could even talk about adding necessary fields to the relevant APIs? Then there would be just one API for defining clusters with on-prem agent-based installs.

Of course the future interactive flow for this tool probably won't use the k8s API as the source of truth (though kcp could make that possible...), but would be more likely to use the REST API. But still, the same basic resources would have to be created, and there are existing translations from CRDs -> rest api resources.

Hopefully this makes more sense.

Reply:
> IIUC, here you're trying to provide a way for a user of this CLI to pass essentially the same information, plus a few additional items that you listed above, which aren't yet covered by CRDs. That user will put the information into one or more text files and then run the CLI.
> 
> This document proposes to use the install-config and a new agent-config as the structure for collecting the info in text files. I'm trying to ask, what if we used the CRDs instead, and filled in any small gaps with annotations?

To be clear, this document is proposing _both_ of those things.

The user can optionally start with install-config and agent-config, then run `openshift-install agent create manifests` to generate the ZTP CRs. Then using agent-config and either CRs generated in the previous step or ones they've created directly themselves, they can run `openshift-installer agent create image` to generate the ISO.

> Or we could even talk about adding necessary fields to the relevant APIs?

It's never going to make sense to e.g. add an API to specify the node 0 address in ZTP, because there isn't a node 0. I do think that everything will end up being represented somehow in the ZTP CRs, because of the way the installer's asset model works.

> Then there would be just one API for defining clusters with on-prem agent-based installs.

I don't think the existence of the optional install-config -> ZTP CRs step takes anything away from anybody who wants to use the ZTP APIs.

> Of course the future interactive flow for this tool probably won't use the k8s API as the source of truth (though kcp could make that possible...),

I keep hearing this... are people talking about a different kcp to https://github.com/kcp-dev/kcp?

> but would be more likely to use the REST API. But still, the same basic resources would have to be created, and there are existing translations from CRDs -> rest api resources.

I guess we could require the user to generate and edit annotations on a bunch of CRs, but that doesn't seem like a very user-friendly way to begin what is supposed to be an interactive install.

Reply:
Thanks for the responses. I'll try to clear up some thoughts from my side.

> It's never going to make sense to e.g. add an API to specify the node 0 address in ZTP, because there isn't a node 0.

Could be. So that's a case where an annotation may be a good fit. It extends the API for a particular use case that isn't broadly applicable. That's a standard pattern for annotations.

> I don't think the existence of the optional install-config -> ZTP CRs step takes anything away from anybody who wants to use the ZTP APIs.

It's just another set of APIs to document, support, lifecycle, etc. And for the user it's another set of APIs to learn, automate against, track changes for over time, etc. It's not the end of the world, but offering one API seems like a better UX than offering two APIs that define the same thing.

> I keep hearing this... are people talking about a different kcp to https://github.com/kcp-dev/kcp?

That's the one. This is future-looking and not central to what I'm asking, but it is interesting! It's a super-simple and cheap way to start a k8s API that mostly just has CRDs. We could run it on node0, add the hive and agent-related CRDs to it, and have assisted-service watch those CRDs like it would in any other k8s control plane. Early on, the kcp folks were even telling a story for compiling kcp into a binary with controllers to make one out-of-the-box artifact; I don't know if that's still something they are pursuing, but it's another possibility.

> I guess we could require the user to generate and edit annotations on a bunch of CRs, but that doesn't seem like a very user-friendly way to begin what is supposed to be an interactive install.

Agreed. I don't think an interactive install involves editing text files at all, right? A UI interactively gets information from the user and creates API resources behind the scenes as it goes along. In this scenario, the format of the cluster definition and related APIs is primarily a concern for the UI developers, not the UI users. Or am I misunderstanding some part of the interactive story?

Reply:
As I understand it, even interactive flow will require some input, since billis expectation is that all the isos are generated in advance, right? So in order to all the nodes to find each other, at least node0 ip should be specified somehow. I agree with @mhrivnak though - lets find a way how we add this input to existing crds, rather than adding new yaml that will duplication (will it?) with existing api (at least partially).  we even can think of the operator configmap cr maybe as a place for those values?

Reply:
Btw, I assume that at some point, we'll generate billi iso in saas (rather than using cli) - how would it work then?

Reply:
> Agreed. I don't think an interactive install involves editing text files at all, right?

Ideally not, but in some places it's unavoidable. e.g. to set up storage multipathing on the root disk, you have to set a kernel command line argument. The only way to do that is to bake it into the image.

Reply:
So to summarise, I see three concerns in this thread.

1. whether extensions to install-config should go into the same install-config.yaml file or a separate agent-config.yaml file.
2. whether all data will be fully represented in the ZTP manifests
3. whether we should support two different input formats and translation from one to the other.

I think the first one is a red herring. I'm not sure anybody really thinks it makes a difference.

For the second, I think it will have to be. As the installer generates the assets for each target, it deletes the assets that were the inputs to that target (to allow you to edit the output without having it overwritten by the next run - the graph works a bit like `make` except that it doesn't check modification times to see if the output is newer than the input, it just deletes the input). So any data represented at the install-config stage will probably have to be represented somewhere in the ZTP CRs as well lest we lose it. Although the graph thing is slightly bananas, this will force us to keep parity over time so it's probably good in the end.

On the third one, I think we need to face the reality that the ZTP CRs are just not very ergonomic to create from scratch in text files - data is distributed in ways that would be odd to a user (partly because it's designed to be able to define multiple clusters in the same namespace, and we only deal with one cluster here), and linked together using k8s API concepts that make perfect sense in the environment they're designed for, but much less so for a bunch of flat input files that will never hit a k8s API. Even our team has had trouble creating them for testing. AIUI, in the environments where ZTP is used people don't seem to write these CRs by hand, but instead generate them from a SiteConfig.

The purpose of the assisted technologies has always been to provide a user-friendly interface, so it was always my assumption that we would need some simpler input format that could generate the ZTP format CRs to then allow users to customise the more esoteric parts, in a similar way to how IPI installs generate the manifests from the install-config in one step and then the bootstrap from the manifests in a second step. We considered both install-config and SiteConfig for this purpose. We chose the former partly because it is designed to describe a single cluster, whereas SiteConfig is designed for multiple clusters, and partly because of Conway's Law, effectively: SiteConfig is controlled by the telco team and it's difficult to justify having the core product depend on outer layers like that, whereas install-config is already part of the OpenShift org.

> It's just another set of APIs to document, support, lifecycle, etc. And for the user it's another set of APIs to learn, automate against, track changes for over time, etc. It's not the end of the world, but offering one API seems like a better UX than offering two APIs that define the same thing.

This is a fair point, and although the idea of having both formats was my proposal, I've tried to keep an open mind about it, because there is indeed _every possibility that this could be a mistake_. We may look back and wish we had never exposed the ZTP format to the user, and instead kept only the install-config interface that is already well documented, familiar to users, has a well-defined lifecycle that is locked to OpenShift releases, &c. I particularly worry that the way we're accepting ZTP data in particular files with particular names (as opposed to throwing all of the k8s objects from one or more files into a k8s API and letting the controllers figure out what you have) might lead to not being interoperable in practice (or rather, with the data being compatible in only one direction) with ZTP deployments through MCE. And while I initially thought that we needed the ZTP stage so that users could add custom manifests, the interface for doing so is so unergonomic (wrapping the custom manifest in a ConfigMap IIRC) that we will probably need a mechanism where we take raw manifests and wrap them into ZTP format on behalf of the user anyway, at which point exposing the ZTP CRs to users might actually be redundant. So this thread is definitely increasing my doubts.

Still, I've been willing to take the risk because we finally have the opportunity to bring the ZTP and installer APIs together and be able to tell a story about how they relate. That seemed worth it, especially given that the ZTP and ReST APIs are similar enough that we are doing most of the work internally already. Perhaps we should delay to a later stage of development when we have a better idea of how the install-config works alone though. I'd be interested to hear from folks on the team who are closer to the implementation.

Thanks for the thread, it has been thought-provoking.